File: Objective-C/Sorting/sorting algorithms/SelectionSort.m
//
//  SelectionSort.m
//  Sorting
//
//  Created by Paul Kraft on 14.04.16.
//  Copyright © 2016 pauljohanneskraft. All rights reserved.
//

#import "SelectionSort.h"

@implementation SelectionSort

+ (void) sort:(NSMutableArray *)array fromLeft:(long)start toRight:(long)end {
    for( long i = start; i <= end; i++) {
        long min = i;
        for( long j = i + 1; j <= end; j++ ) {
            if( [array objectAtIndex:j] < [array objectAtIndex:min])
                min = j;
        }
        [self swap:array index1:i index2:min];
    }
}

@end

File: Swift/sorting algorithms/SelectionSort.swift
//
//  SelectionSort.swift
//  Sort
//
//  Created by Paul Kraft on 21.12.15.
//  Copyright © 2015 Paul Kraft. All rights reserved.
//

func selectionSort<T: Comparable>(array: [T]) -> (name: String, array: [T]) {
    var sorted = array
    selectionSort(&sorted)
    return ("SelectionSort", sorted)
}

func selectionSort<T: Comparable>(inout array: [T]) {
    selectionSort(&array, 0..<array.count)
}

func selectionSort<T: Comparable>(inout array: [T], _ range: Range<Int>) {
    for left in range {
        var min = left
        for i in left+1..<array.count {
            if(array[i] < array[min]) { min = i }
        }
        array[left] <-> array[min]
    }
}